core/App/ffi_operation.cpp

```diff
@@ -1535,4 +1535,82 @@ extern "C"
         return EH_OK;
     }
 
+    /**
+     * @brief Generate Hmac (SHA-256) with given apikey(encrypted), cmk and payload
+     * @param payload : Pass in the cmk, apikey(encrypted) and payload in the form of JSON string
+                {
+                    cmk : a base64 string,
+                    apikey: a base64 string
+                    payload : a string,
+                }
+     * @return [string] json string
+                {
+                    code: int,
+                    message: string,
+                    result: {
+                        hmac: string,
+                    }
+                }
+     */
+    uint32_t ffi_generateHmac(JsonObj payloadJson, char *respJson)
+    {
+        ehsm_status_t ret;
+        // input params
+        ehsm_keyblob_t *cmk = NULL;
+        ehsm_data_t *apikey = NULL;
+        ehsm_data_t *payload = NULL;
+        // immediate vars
+        size_t payload_size;
+        std::string payload_str;
+        // output params
+        ehsm_data_t *hmac = NULL;
+        std::string hmac_str;
+        RetJsonObj retJsonObj;
+
+        // 0. prepare data
+        JSON2STRUCT(payloadJson, cmk);
+        JSON2STRUCT(payloadJson, apikey);
+
+        payload_str = payloadJson.readData_string("payload");
```

Comment:
check if null?

Reply:
Thanks for reviewing! The validation of `cmk` and `apikey` should be added in the latest commit, please chceck.